selfservice/flow/registration/handler.go

```diff
@@ -318,29 +318,79 @@ type createBrowserRegistrationFlow struct {
 //	  303: emptyResponse
 //	  default: errorGeneric
 func (h *Handler) createBrowserRegistrationFlow(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	ctx := r.Context()
+
 	a, err := h.NewRegistrationFlow(w, r, flow.TypeBrowser)
 	if err != nil {
-		h.d.SelfServiceErrorManager().Forward(r.Context(), w, r, err)
+		h.d.SelfServiceErrorManager().Forward(ctx, w, r, err)
 		return
 	}
 
-	if sess, err := h.d.SessionManager().FetchFromRequest(r.Context(), r); err == nil {
-		if r.URL.Query().Has("login_challenge") {
-			logoutUrl := urlx.AppendPaths(h.d.Config().SelfPublicURL(r.Context()), logout.RouteSubmitFlow)
-			self := urlx.CopyWithQuery(
-				urlx.AppendPaths(h.d.Config().SelfPublicURL(r.Context()), RouteInitBrowserFlow),
-				r.URL.Query(),
-			).String()
-
-			http.Redirect(
-				w,
-				r,
-				urlx.CopyWithQuery(logoutUrl, url.Values{
-					"token":     {sess.LogoutToken},
-					"return_to": {self},
-				}).String(),
-				http.StatusFound,
-			)
+	var (
+		hydraLoginRequest   *hydraclientgo.OAuth2LoginRequest
+		hydraLoginChallenge sqlxx.NullString
+	)
+	if r.URL.Query().Has("login_challenge") {
+		var err error
+		hydraLoginChallenge, err = hydra.GetLoginChallengeID(h.d.Config(), r)
+		if err != nil {
+			h.d.SelfServiceErrorManager().Forward(ctx, w, r, err)
+			return
+		}
+
+		hydraLoginRequest, err = h.d.Hydra().GetLoginRequest(ctx, hydraLoginChallenge.String())
+		if err != nil {
+			h.d.SelfServiceErrorManager().Forward(ctx, w, r, err)
+			return
+		}
+
+		// on OAuth2 flows, we need to use the RequestURL
+		// as the ReturnTo URL.
+		// This is because a user might want to switch between
+		// different flows, such as login to registration and login to recovery.
+		// After completing a complex flow, such as recovery, we want the user
+		// to be redirected back to the original OAuth2 login flow.
+		if hydraLoginRequest.RequestUrl != "" && h.d.Config().OAuth2ProviderOverrideReturnTo(r.Context()) {
+			q := r.URL.Query()
+			// replace the return_to query parameter
+			q.Set("return_to", hydraLoginRequest.RequestUrl)
+			r.URL.RawQuery = q.Encode()
+		}
+	}
+
+	if sess, err := h.d.SessionManager().FetchFromRequest(ctx, r); err == nil {
+
+		if hydraLoginRequest != nil {
+			// hydra indicates that we cannot skip the login request
+			// so we must perform the login flow.
+			if !hydraLoginRequest.GetSkip() {
```

Comment:
I think skip is used purely for the consent. For forcing the login, we need to look for `prompt=login`. Prompt can have space-delimited values, so `prompt=login registration` would be valid.

Reply:
what should we then do if `skip=false`? In this case the user cannot go back to the client, on the Login flow we usually do `refresh=true` here. So this is why I redirect to login and have the login logic handle what should happen to the OAuth flow. 

In this case, the user is re-prompted to login or "confirm it's you".